contracts/interfaces/IV4Router.sol

```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import "forge-std/console.sol";
+import {Hooks} from "@uniswap/v4-core/contracts/libraries/Hooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/contracts/interfaces/IPoolManager.sol";
+import {BalanceDelta} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/contracts/interfaces/IHooks.sol";
+
+/// @title UniswapV4Routing
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+interface IV4Router {
+    error NotPoolManager();
+    error InvalidSwapType();
+    error TooLittleReceived();
+    error TooMuchRequested();
+
+    struct SwapInfo {
+        SwapType swapType;
+        address msgSender;
+        bytes params;
+    }
+
+    struct PathKey {
+        Currency tradeCurrency;
+        uint24 fee;
+        int24 tickSpacing;
+        IHooks hooks;
+    }
+
+    struct ExactInputSingleParams {
+        PoolKey poolKey;
+        bool zeroForOne;
+        address recipient;
+        uint128 amountIn;
+        uint128 amountOutMinimum;
+        uint160 sqrtPriceLimitX96;
```

Comment:
Why are both `sqrtPriceLimitX96` and `amountOutMinimum` needed? Dont they both achieve slippage protection?

Reply:
In fact I'm pretty sure `amountOutMinimum` is unused?

Reply:
ohh it was actually on my TODO to see how UR/router-sdk deals with slippage. amountOutMinimum is typically used from our UI. The only time Ive used `sqrtPriceLimitX96` is with TWAMM which needs it (but not through a router).`sqrtPriceLimitX96` is less about slippage and more about halting a swap midway through. Noah mentioned he was glad TWAMM needed it bc they were on the fence about adding the feature in. Anyway it's there for people if they need it, but it's not used much...

Reply:
I think it could be cool to keep it for others to use if they want (we wouldn't be using any of the swapSingle in UR anyway, which is the only place it's exposed) But if unused internal functions contribute to UR bytecode (can experiment with that) happy to take out

Reply:
yeah im pro keeping it for now, other people could well use it even if we dont